metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/MainRewards.scala

```diff
@@ -5,83 +5,127 @@ import cats.syntax.all._
+  def make[F[_] : Async : SecurityProvider]: Rewards[F, CurrencySnapshotStateProof, CurrencyIncrementalSnapshot, CurrencySnapshotEvent] = {
     (
-      lastArtifact: Signed[CurrencyIncrementalSnapshot],
-      lastBalances: SortedMap[Address, Balance],
-      _: SortedSet[Signed[Transaction]],
-      consensusTrigger: ConsensusTrigger
-      , _: Set[CurrencySnapshotEvent],
+      lastArtifact        : Signed[CurrencyIncrementalSnapshot],
+      lastBalances        : SortedMap[Address, Balance],
+      _                   : SortedSet[Signed[Transaction]],
+      consensusTrigger    : ConsensusTrigger,
+      _                   : Set[CurrencySnapshotEvent],
       maybeCalculatedState: Option[DataCalculatedState]
     ) => {
-      consensusTrigger match {
-        case trigger.EventTrigger => SortedSet.empty[transaction.RewardTransaction].pure[F]
-        case trigger.TimeTrigger =>
-          val currentEpochProgress = lastArtifact.epochProgress.value.value + 1
-          val epochProgressModulus = currentEpochProgress % EPOCH_PROGRESS_1_DAY
-          if (epochProgressModulus > 2) {
-            SortedSet.empty[RewardTransaction].pure[F]
-          } else {
-            maybeCalculatedState
-              .foldM(SortedSet.empty[transaction.RewardTransaction]) { (_, calculatedState) =>
-                logger.info("Starting the rewards...")
-                for {
-                  (l0ValidatorNodes, l1ValidatorNodes) <- getValidatorNodes
-                  _ = logInitialRewardDistribution(epochProgressModulus)
-                } yield buildRewards(calculatedState.asInstanceOf[CheckInDataCalculatedState], currentEpochProgress, lastBalances, l0ValidatorNodes, l1ValidatorNodes)
-              }
+      def distributeRewards(
+        lastArtifact        : Signed[CurrencyIncrementalSnapshot],
+        lastBalances        : SortedMap[Address, Balance],
+        maybeCalculatedState: Option[DataCalculatedState]
+      ): F[SortedSet[RewardTransaction]] = {
+        val currentEpochProgress = lastArtifact.epochProgress.value.value + 1
+        val epochProgressModulus = currentEpochProgress % EpochProgress1Day
+        if (epochProgressModulus > 2) {
+          SortedSet.empty[RewardTransaction].pure[F]
+        } else {
+          maybeCalculatedState.foldLeftM(SortedSet.empty[RewardTransaction]) { (_, calculatedState) =>
+            for {
+              _ <- logger.info("Starting the rewards...")
+              (l0ValidatorNodes, l1ValidatorNodes) <- getValidatorNodes
+              _ <- logInitialRewardDistribution(epochProgressModulus)
+              checkInCalculatedState: CheckInDataCalculatedState = calculatedState.asInstanceOf[CheckInDataCalculatedState]
+              rewards <- buildRewards(checkInCalculatedState, currentEpochProgress, lastBalances, l0ValidatorNodes, l1ValidatorNodes)
+            } yield rewards
           }
```

Comment:
<img width="2537" alt="Captura de Tela 2023-12-01 às 11 00 13" src="https://github.com/Constellation-Labs/dor-metagraph/assets/22091534/1b166124-bb50-493a-873a-9e4a205a519b">

Reply:
@sdavidp look at prints above

Reply:
Remember I warned of the dangers in using `asInstanceOf` ?
I suspect that is the problem.
I would make this change; hopefully it will print before blowing up and tell us what we've got.
```
_ <- logger.warn(s"*** calculatedState is $calculatedState ***")  // ADD THIS LINE
checkInCalculatedState: CheckInDataCalculatedState = calculatedState.asInstanceOf[CheckInDataCalculatedState]
```

Reply:
For the record, the error was in my suggested refactor: You can't use `whenA` and `unlessA` on `Option`. :p